enhancements/installer/aws-cross-account-dns-zone.md

```diff
@@ -162,9 +160,55 @@ platform:
     hostedZone: Z00147933I3NWOQ6M4699
     hostedZoneRole: arn:aws:iam::<account-a>:role/<role-name>
 ```
-Please see Open Questions for further discussion of the install config.
 
+### API: ExternalDNS Operator
+
+[ExternalDNS Operator](https://github.com/openshift/external-dns-operator) allows you to deploy and manage [ExternalDNS](https://github.com/kubernetes-sigs/external-dns),
+a cluster-internal component which makes Kubernetes resources discoverable through public DNS servers. This enhancement
+extends the ExternalDNS Operator to support cross-account DNS zones in AWS. We extend the`ExternalDNS` API object by adding
+the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
+Refer to [`externaldns_types.go`](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)
+for the existing API structure of the `ExternalDNS` object.
+
+In addition to adding `RoleARN`, the `Credentials` field was updated to be optional. Prior to this change, the `Credentials`
```

Comment:
>  the `Credentials` field was updated to be optional

I think we have to consult the API team on that change. It's a tricky thing, there are a lot of caveats: how the object is stored in etcd, API calls vs manifests, server-side apply, etc. My fear is that even a webhook may be not enough here.

Reply:
I don't think I'm following completely. Could you elaborate on your concern here? What scenarios are you envisioning `optional` being an issue? 

Is your concern with just `Credentials` being optional? Or is your concern with the updating `Credentials` from `required` to `optional`?

Reply:
> Or is your concern with the updating Credentials from required to optional?

Yes.

I think I came up with an example which may cause issues if we don't bump a new ExternalDNS api version.
Imagine a client application calling ExternalDNS API:
```go
func getAWSCredentialsName() string { 
  extdns := &extdnsv1beta1.ExternalDNS{}
  ocClient.Get(extdns)
  if extdns.spec.provider.type == "AWS" && extdns.spec.provider.aws != nil {
    return extdns.spec.provider.aws.Credentials
  }
  return ""
}
```
This code will become faulty.

Reply:
Ah okay I see. The assumption they were making was that `extdns.Spec.Provider.AWS.Credentials` was always not-nil because it was required.

So sound like we need to bump the version of External DNS operator API. Looking back at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility, we violate `which fields are required and which are not`.

So...how do we go about bumping the API? Can we do a `v1beta2` similar to what Gateway API has [done](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1alpha2)?

Reply:
> So sound like we need to bump the version of External DNS operator API. Looking back at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility, we violate which fields are required and which are not.

Yes.

> Can we do a v1beta2 similar to what Gateway API has [done](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1alpha2)?

Yeah, I think `v1beta2` should be fine.

> So...how do we go about bumping the API?

It should be a new api version with a conversion webhook to handle a situation when `ExternalDNSAWSProviderOptions` is empty. The conversion webhook would force us to change the install mode for the operator to `AllNamespaces`. I did a similar exercise when `v1` version of ALBO was released ([PR](https://github.com/openshift/aws-load-balancer-operator/pull/93)). 

Overall it's a pity we have to do all this just to a add a new API field. We didn't think about setting the field optional from [the introduction of the API](https://github.com/openshift/external-dns-operator/blame/main/api/v1alpha1/externaldns_types.go#L258) and now it strikes back.

Maybe API team may come up with some tricks of how to approach this differently (without a new API version). For instance, a new field or a new union, I gave myself 30 mins to think of it but didn't come with any decent idea.

Reply:
Resolving thread as we discussed a way forward in https://redhat-internal.slack.com/archives/CE4L0F143/p1691007201050019?thread_ts=1690552114.405509&cid=CE4L0F143

For more context, decision is documented in https://github.com/openshift/external-dns-operator/pull/195/